### ISSUE TITLE
Multiuser chat bot with include_callback_query_chat_id mixes up different user's responses

### ISSUE THREAD

**Issue by eliastsegaye** [nickoala/telepot#346]
Hello!
I wrote a quiz bot , and used the "include_callback_query_chat_id" to handle callback queries, however it seems that the the bot handles different users if they were the same user.  Am I getting it wrong somewhere?

class MessageCounter(telepot.helper.ChatHandler):
	user_state={'ms':0,'ss':0}
	selected_batch=""
	current_question=0
	waiting=0
	random_order=[]
	is_random=0
	def __init__(self, *args, **kwargs):
		super(MessageCounter, self).__init__(*args, **kwargs)
		self._count = 0
		
	def on_callback_query(self, msg):
		start_questions=ReplyKeyboardMarkup(keyboard=[[KeyboardButton(text='Start Questions---->')]])
		query_id, from_id, query_data = telepot.glance(msg, flavor='callback_query')
		#response arrives from the menu item sent earlier on
		if (self.user_state['ms']==0 and self.user_state['ss']==1):
			#change state variable to current state
			self.user_state['ms']=2
			self.user_state['ss']=0
			
			#load question
			question_to_load=query_data+".json"
			question_name=self.load_question(question_to_load)
			
			#initialize the random_order array if randomize flag is set to one
			if self.is_random==1:
				self.randomize_order(len(self.selected_batch))
			
			#todo change to menu
			self.sender.sendMessage("Loaded question batch: \n" +question_name+". ",reply_markup=start_questions)
			#self.sender.sendMessage(str(self.selected_batch))
			pass
			
			#TODO send start question 		
				
		print query_data
	def on_chat_message(self, msg):
		try:
			self._count += 1
			#self.sender.sendMessage(self._count)
			choices=ReplyKeyboardMarkup(keyboard=[[KeyboardButton(text='A')],[KeyboardButton(text='B')],[KeyboardButton(text='C')],[KeyboardButton(text='D')],])
			next_question=ReplyKeyboardMarkup(keyboard=[[KeyboardButton(text='Next Question ---->')]])
			start_questions=ReplyKeyboardMarkup(keyboard=[[KeyboardButton(text='Start Questions---->')]])
			
			#handle start and menu commands
			if ((msg['text'].lower())=="/start" or (msg['text'].lower())=="/menu"):
				self.sender.sendMessage("Welcome you wonderful person, you!")
				self.user_state['ms']=0
				self.user_state['ss']=0
				#return 1
			
			#handle random order
			if ((msg['text'].lower())=="/random"):
				self.is_random=1
				self.sender.sendMessage("Question order is set to random!")
				self.user_state['ms']=0
				self.user_state['ss']=0
			
			#handle random order
			if ((msg['text'].lower())=="/straight"):
				self.is_random=0
				self.sender.sendMessage("Question order is set to straight!")
				self.user_state['ms']=0
				self.user_state['ss']=0
				
			
			#questions other than the first question	
			if self.user_state['ms']==2 and self.user_state['ss']==2:
				max_questions=len(self.selected_batch)
				if self.current_question>=max_questions-1:
					self.sender.sendMessage("This batch of questions is done. Choose another batch")
					#todo send main menu batch
					self.user_state['ms']=0
					self.user_state['ss']=0
					pass
				else:
					if self.is_random==1:
						print "random:" + str(self.current_question)
						self.sender.sendMessage(self.selected_batch[self.random_order[self.current_question]]['question'],reply_markup=choices)
					else:
						self.sender.sendMessage(self.selected_batch[self.current_question]['question'],reply_markup=choices)
					self.user_state['ss']=1
					self.waiting=1
					pass
			
			
			#answer of the  questions	
			if self.user_state['ms']==2 and self.user_state['ss']==1:
				if self.waiting==1:
					self.waiting=0
				else:
					if self.is_random==1:
						if self.selected_batch[self.random_order[self.current_question]]['answer'].lower()==msg['text'].lower():
							self.sender.sendMessage("The answer is correct!",reply_markup=next_question)
							
						else:
							self.sender.sendMessage("The answer is not correct",reply_markup=next_question)
						
						description=self.selected_batch[self.random_order[self.current_question]]['description']
						if description!="":
							self.sender.sendMessage(description)
					else:
						if self.selected_batch[self.current_question]['answer'].lower()==msg['text'].lower():
							self.sender.sendMessage("The answer is correct!",reply_markup=next_question)
							
						else:
							self.sender.sendMessage("The answer is not correct",reply_markup=next_question)
						
						description=self.selected_batch[self.current_question]['description']
						if description!="":
							self.sender.sendMessage(description)
						
					self.current_question=self.current_question+1
					self.user_state['ss']=2
					pass
				#todo send a next button		
			
			#START question state, send first question
			if self.user_state['ms']==2 and self.user_state['ss']==0:
				self.current_question=0
				#todo send first question
				print self.selected_batch
				
				#if the order of quesitons is randomized, use random order array to reorder questions
				if self.is_random==1:
					print "random:" + str(self.current_question)
					self.sender.sendMessage(str(self.selected_batch[self.random_order[self.current_question]]['question']),reply_markup=choices)
				else:
					self.sender.sendMessage(str(self.selected_batch[self.current_question]['question']),reply_markup=choices)
				self.user_state['ss']=1			
				pass
				
			#reply to initial menu (question batch selection)
			if (self.user_state['ms']==0 and self.user_state['ss']==1):
			#Kindly inform user to select using the menu provided
				self.sender.sendMessage( 'Hi, Please use the menu provided to select question!')
				
			#menu state
			#initial menu state(menu not displayed to user yet)
			if (self.user_state['ms']==0 and self.user_state['ss']==0):
				#set menu state to 1 to signify message sent to user
				self.user_state['ss']=1
				keyboard_buttons=[]
				counter=1
				menu_item=""
				for question in self.list_question_batch("."):
					menu_item=(str(counter)+".) "+ question.replace(".json",""))
					keyboard_buttons.append([InlineKeyboardButton(text=menu_item, callback_data=str(question.replace(".json","")))])
					counter+=1
				keyboard = InlineKeyboardMarkup(inline_keyboard=keyboard_buttons)
				self.sender.sendMessage( 'Select Question Batch', reply_markup=keyboard)
				pass
				#TODO change to initial menu			
				
			if msg['chat']['type']=='document':
				self.sendMessage("document received")
				
		except Exception as e:
			print e
			print traceback.print_exc()
			self.sender.sendMessage("Sorry, an error occured! \n Try sending me the appropriate question number. ")
	
	def load_question(self,question_file):
		json_question=open(question_file).read()
		questions=json.loads(json_question)
		self.selected_batch=questions
		return str(questions[0]["question_info"])
		
	
			
	def list_question_batch(self,questions_path):
		#Takes a folder as arguement and returns the list of json files therein'''
		#default_path='/questions'
		question_batches=[]
		for file in os.listdir(questions_path):
			if file.endswith('.json'):
				question_batches.append(file)
			#
			#	question_batches.append(os.path.join(questions_path,file))
		return question_batches
	
	#given the length of an array, returns a random ordering of the array
	def randomize_order(self,ranges):
		random_order=[]
		for i in range(0,ranges-1):
			x=randint(0,ranges-1)
			while x in random_order:
				x=randint(0,ranges-1)
			random_order.append(x)
		self.random_order=random_order
		return random_order
		

		
	
TOKEN = "451887625:AAF1KFrPZK03LQIAQT70at8XZKPt2JwkyHw" # get app token from command-line
#TOKEN="377784959:AAHWnZ3Nts7GpElqubS90F39I2C_pth7f6I"
bot = telepot.DelegatorBot(TOKEN, [
    include_callback_query_chat_id(
        pave_event_space())(
            per_chat_id(types='all'), create_open, MessageCounter, timeout=600000),
])

**Comment by das7pad** [comment]
Hello @eliastsegaye,

please revoke the api keys for your bots. They should remain private as everybody can use them now.

Use the `/revoke` command in the chat with `@BotFather`. Just editing the Issue does not help.

---

Could you rephrase this part?

 > the bot handles different users if they were the same user

**Comment by eliastsegaye** [comment]
Hi das7pad, the not is supposed to load questions from a json file and ask users. When different users start using it, it takes the response of one user and run responds to another user. For example, user 1 could be answering question 1 whereas user 2 is selecting which questions to load. In this case, when user 2 sends a message, the bot assumes that it was user 1 interacting and changes user 1's state variables. Is this clearer?
